### ISSUE TITLE
STL headers should require C++ and forbid C

### ISSUE THREAD

**Issue by StephanTLavavej** [microsoft/STL#1141]
Until fairly recently, the STL's build contained C translation units, so some of our internal headers had to tolerate being compiled as C. Now that our build contains only C++ TUs, we should consider changing our headers to require C++. (This was prompted by discovering that an internal team had been including `<cassert>` in C mode. They've fixed their code.)

iso646.h
---
https://github.com/microsoft/STL/blob/b74b6188b233cce0647950d7fd6bbbb3bd3744f0/stl/inc/iso646.h#L6-L13

This header is special. It's a C header that lived in the STL's directory and (mistakenly?) wasn't moved to the UCRT. I believe we should:
* Stop including `<yvals_core.h>`
* Replace `#if _STL_COMPILER_PREPROCESSOR` with equivalent machinery
* Keep the `__cplusplus` guard
* ~~Investigate moving this to the UCRT properly~~ (Edit: not worth the effort)

xkeycheck.h
---
https://github.com/microsoft/STL/blob/e745bad3b1d05b5b19ec652d68abb37865ffa454/stl/inc/xkeycheck.h#L6-L15

On 2019-08-21, Microsoft-internal MSVC-PR-198458 changed this header to have no effect in C mode.

I think we should:

* Remove the `__cplusplus` guard
* Rely on `<yvals_core.h>` enforcing C++ mode

yvals_core.h
---
https://github.com/microsoft/STL/blob/b74b6188b233cce0647950d7fd6bbbb3bd3744f0/stl/inc/yvals_core.h#L1237-L1247

This is the STL's central internal header. I think we should:

* Remove the C definitions here
* Add an `#error` elsewhere in the file, probably early, forbidding it from being compiled as C (like how we forbid old compilers)

yvals.h
---
https://github.com/microsoft/STL/blob/b74b6188b233cce0647950d7fd6bbbb3bd3744f0/stl/inc/yvals.h#L133-L153
https://github.com/microsoft/STL/blob/b74b6188b233cce0647950d7fd6bbbb3bd3744f0/stl/inc/yvals.h#L325-L501

This is the STL's "runtime support" internal header. I think we should:
* Remove the `__cplusplus` guards here
* Rely on `<yvals_core.h>` enforcing C++ mode

**Comment by BillyONeal** [comment]
I agree with all your bullets.

**Comment by CaseyCarter** [comment]
Mutual inclusion of `xkeycheck.h` <=> `yvals_core.h` is intentional to ensure that `xkeycheck.h` has the proper value for `_STL_COMPILER_PREPROCESSOR` even if it's included before/without including `yvals_core.h`. If we don't want it to be possible to include `xkeycheck.h` without including `yvals_core.h`, we should fuse the headers.

Honestly I'd prefer to dump `xkeycheck.h` and request a `#pragma` to instruct the preprocessor to gripe at the user about keyword definitions. `xkeycheck.h` is a poor solution to this problem:
```c++
#include <ciso646>
#define private public
#include <vector> // Oops: we don't recheck
```
or
```c++
#define likely potato
#include <vector> // Oops: we can't defend object-like "likely" without breaking the common function-like "likely()"
```
and I consequently don't think it provides good return on investment. Other STLs get by just fine without a solution for this problem - they haven't exactly been banging on our door demanding `<xkeycheck.h>` since we open-sourced ;)

**Comment by BillyONeal** [comment]
We've been asking compilers to forbid it since the dawn of time, so I don't think such a pragma will be forthcoming.

This isn't about enforcement for customers' sake, it's about the number of 'impossible' bug reports we've gotten from people doing things like defining `return` and then complaining that parts of the library aren't constexpr.

**Comment by StephanTLavavej** [comment]
`xkeycheck.h` circularity was fixed by #1692, I'll update the issue description.
